modules/pam_pwhistory/pam_pwhistory.c

```diff
@@ -105,14 +108,184 @@ parse_option (pam_handle_t *pamh, const char *argv, options_t *options)
     pam_syslog (pamh, LOG_ERR, "pam_pwhistory: unknown option: %s", argv);
 }
 
+static int
+run_save_helper(pam_handle_t *pamh, const char *user,
+		int howmany, int debug)
+{
+  int retval, child;
+  struct sigaction newsa, oldsa;
+
+  memset(&newsa, '\0', sizeof(newsa));
+  newsa.sa_handler = SIG_DFL;
+  sigaction(SIGCHLD, &newsa, &oldsa);
+
+  child = fork();
+  if (child == 0)
+    {
+      static char *envp[] = { NULL };
+      char *args[] = { NULL, NULL, NULL, NULL, NULL, NULL };
+
+      if (pam_modutil_sanitize_helper_fds(pamh, PAM_MODUTIL_PIPE_FD,
+          PAM_MODUTIL_PIPE_FD,
+          PAM_MODUTIL_PIPE_FD) < 0)
+        {
+          _exit(PAM_SYSTEM_ERR);
+        }
+
+      /* exec binary helper */
+      DIAG_PUSH_IGNORE_CAST_QUAL;
+      args[0] = (char *)PWHISTORY_HELPER;
+      args[1] = (char *)"save";
+      args[2] = (char *)user;
+      DIAG_POP_IGNORE_CAST_QUAL;
+      if (asprintf(&args[3], "%d", howmany) < 0)
+        {
+          pam_syslog(pamh, LOG_ERR, "asprintf on args[3] failed");
+          _exit(PAM_SYSTEM_ERR);
+        }
+      if (asprintf(&args[4], "%d", debug) < 0)
+        {
+          pam_syslog(pamh, LOG_ERR, "asprintf on args[4] failed");
+          _exit(PAM_SYSTEM_ERR);
+        }
```

Comment:
Error diagnostics here is too verbose.  Something more concise like
```
if (asprintf(&args[3], "%d", howmany) < 0 ||
    asprintf(&args[4], "%d", debug) < 0)
  {
    pam_syslog(pamh, LOG_ERR, "asprintf: %m");
    _exit(PAM_SYSTEM_ERR);
  }
```
should be enough.